File: VB/Storages/XpoFileStorage.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Data.Filtering
Imports DevExpress.Xpf.Reports.UserDesigner
Imports DevExpress.Xpf.Reports.UserDesigner.Native
Imports DevExpress.Xpo
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Windows

Namespace WpfApplication38
	Public Class XpoFileStorage
		Implements IReportFileStorage
		 Private items_Renamed As XPCollection(Of StorageItem)
		 Public Sub New(ByVal session As UnitOfWork)
			 MyBase.New()
			 items_Renamed = New XPCollection(Of StorageItem)(session)
		 End Sub
        Public Function GetErrorMessage(ByVal exception As Exception) As String Implements IReportFileStorage.GetErrorMessage
            Return ExceptionHelper.GetInnerErrorMessage(exception)
        End Function

        Public Function Load(ByVal filePath As String) As DevExpress.XtraReports.UI.XtraReport Implements IReportFileStorage.Load
            ' Get a StorageItem containing the report.
            Dim item As StorageItem = FindItem(filePath)
            If item IsNot Nothing Then
                Return XtraReport.FromStream(New MemoryStream(item.Layout), True)
            Else
                Dim report As New XtraReport()
                report.Bands.Add(New DetailBand())
                Return report
            End If
        End Function

        Public Sub Save(ByVal filePath As String, ByVal report As DevExpress.XtraReports.UI.XtraReport) Implements IReportFileStorage.Save
            TypeDescriptor.GetProperties(GetType(XtraReport))("DisplayName").SetValue(report, filePath)
            SetData(report, filePath)
        End Sub

        Public Function ShowOpenDialog(ByVal designer As DevExpress.Xpf.Reports.UserDesigner.Native.IReportDesignerUI) As String Implements IReportFileStorage.ShowOpenDialog
            Dim form As StorageEditorForm = CreateForm()
            form.textBox1.IsEnabled = False
            Dim result? As Boolean = form.ShowDialog()
            If result.HasValue AndAlso result.Value Then
                Return form.textBox1.Text
            Else
                Return String.Empty
            End If
        End Function

        Public Function ShowSaveAsDialog(ByVal filePath As String, ByVal reportTitle As String, ByVal designer As DevExpress.Xpf.Reports.UserDesigner.Native.IReportDesignerUI) As String Implements IReportFileStorage.ShowSaveAsDialog
            Dim form As StorageEditorForm = CreateForm()
            form.textBox1.Text = reportTitle
            form.listBox1.IsEnabled = False
            ' Show the save dialog to get a URL for a new report.
            Dim result? As Boolean = form.ShowDialog()
            If result.HasValue AndAlso result.Value Then
                Dim url As String = form.textBox1.Text
                If (Not String.IsNullOrEmpty(url)) AndAlso (Not form.listBox1.Items.Contains(url)) Then

                    Return url
                Else
                    MessageBox.Show("Incorrect report name", "Error", MessageBoxButton.OKCancel, MessageBoxImage.Error)
                End If
            End If
            Return String.Empty
        End Function
		Public Sub SetData(ByVal report As XtraReport, ByVal url As String)
			' Write the report to a corresponding StorageItem.
			' If a StorageItem with a specified Url property value does not exist, create a new one.
			Dim item As StorageItem = FindItem(url)
			If item IsNot Nothing Then
				item.Layout = GetBuffer(report)
			Else
				item = New StorageItem(Session)
				item.Url = url
				Session.CommitChanges()

				report.Extensions("StorageID") = item.Oid.ToString()
				item.Layout = GetBuffer(report)
			End If
			Session.CommitChanges()
			items_Renamed.Reload()
		End Sub
		Private Function GetBuffer(ByVal report As XtraReport) As Byte()
			Using stream As New MemoryStream()
				report.SaveLayout(stream)
				Return stream.ToArray()
			End Using
		End Function
		Private Function CreateForm() As StorageEditorForm
			Dim form As New StorageEditorForm()
			For Each item As String In GetUrls()
				form.listBox1.Items.Add(item)
			Next item
			Return form
		End Function
		Private Function GetUrls() As String()
			Return GetUrlsCore(Nothing).ToArray()
		End Function
		Private Function GetUrlsCore(ByVal method As Predicate(Of String)) As List(Of String)
			Dim list As New List(Of String)()
			For Each item As StorageItem In Items
				If method Is Nothing OrElse method(item.Oid.ToString()) Then
					list.Add(item.Url)
				End If
			Next item
			Return list
		End Function
		Private ReadOnly Property Session() As UnitOfWork
			Get
				Return CType(items_Renamed.Session, UnitOfWork)
			End Get
		End Property
		Public Function FindItem(ByVal name As String) As StorageItem
			Return Session.FindObject(Of StorageItem)(New BinaryOperator("Url", name))
		End Function
		Public ReadOnly Property Items() As XPCollection(Of StorageItem)
			Get
				Return items_Renamed
			End Get
		End Property


	End Class

	Public Class StorageItem
		Inherits XPObject
		Private url_Renamed As String
		Private layout_Renamed() As Byte = Nothing
		Public Property Url() As String
			Get
				Return url_Renamed
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Url", url_Renamed, value)
			End Set
		End Property
		Public Property Layout() As Byte()
			Get
				Return layout_Renamed
			End Get
			Set(ByVal value As Byte())
				SetPropertyValue("Layout", layout_Renamed, value)
			End Set
		End Property
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
	End Class
End Namespace

File: CS/WpfApplication38/Storages/XpoFileStorage.cs
using DevExpress.Data.Filtering;
using DevExpress.Xpf.Reports.UserDesigner;
using DevExpress.Xpf.Reports.UserDesigner.Native;
using DevExpress.Xpo;
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows;

namespace WpfApplication38
{
    public class XpoFileStorage : IReportFileStorage
    {
         XPCollection<StorageItem> items;
         public XpoFileStorage(UnitOfWork session)
             : base()
         {
             items = new XPCollection<StorageItem>(session);
        }
        public string GetErrorMessage(Exception exception)
        {
            return ExceptionHelper.GetInnerErrorMessage(exception);
        }

        public DevExpress.XtraReports.UI.XtraReport Load(string filePath)
        {
            // Get a StorageItem containing the report.
            StorageItem item = FindItem(filePath);
            if (item != null)
                return XtraReport.FromStream(new MemoryStream(item.Layout), true);
            else
            {
                XtraReport report = new XtraReport();
                report.Bands.Add(new DetailBand());
                return report;
            }
        }

        public void Save(string filePath, DevExpress.XtraReports.UI.XtraReport report)
        {
            TypeDescriptor.GetProperties(typeof(XtraReport))["DisplayName"].SetValue(report, filePath);
            SetData(report, filePath);
        }

        public string ShowOpenDialog(DevExpress.Xpf.Reports.UserDesigner.Native.IReportDesignerUI designer)
        {
            StorageEditorForm form = CreateForm();
            form.textBox1.IsEnabled = false;
            bool? result = form.ShowDialog();
            if (result.HasValue && result.Value)
                return form.textBox1.Text;
            else return string.Empty;
        }

        public string ShowSaveAsDialog(string filePath, string reportTitle, DevExpress.Xpf.Reports.UserDesigner.Native.IReportDesignerUI designer)
        {
            StorageEditorForm form = CreateForm();
            form.textBox1.Text = reportTitle;
            form.listBox1.IsEnabled = false;
            // Show the save dialog to get a URL for a new report.
            bool? result = form.ShowDialog();
            if (result.HasValue && result.Value)
            {
                string url = form.textBox1.Text;
                if (!string.IsNullOrEmpty(url) && !form.listBox1.Items.Contains(url))
                {

                    return url;
                }
                else
                {
                    MessageBox.Show("Incorrect report name", "Error", MessageBoxButton.OKCancel, MessageBoxImage.Error);
                }
            }
            return string.Empty;
        }
        public void SetData(XtraReport report, string url)
        {
            // Write the report to a corresponding StorageItem.
            // If a StorageItem with a specified Url property value does not exist, create a new one.
            StorageItem item = FindItem(url);
            if (item != null)
                item.Layout = GetBuffer(report);
            else
            {
                item = new StorageItem(Session);
                item.Url = url;
                Session.CommitChanges();

                report.Extensions["StorageID"] = item.Oid.ToString();
                item.Layout = GetBuffer(report);
            }
            Session.CommitChanges();
            items.Reload();
        }
        byte[] GetBuffer(XtraReport report)
        {
            using (MemoryStream stream = new MemoryStream())
            {
                report.SaveLayout(stream);
                return stream.ToArray();
            }
        }
        StorageEditorForm CreateForm()
        {
            StorageEditorForm form = new StorageEditorForm();
            foreach (string item in GetUrls())
                form.listBox1.Items.Add(item);
            return form;
        }
        string[] GetUrls()
        {
            return GetUrlsCore(null).ToArray();
        }
        List<string> GetUrlsCore(Predicate<string> method)
        {
            List<string> list = new List<string>();
            foreach (StorageItem item in Items)
                if (method == null || method(item.Oid.ToString()))
                    list.Add(item.Url);
            return list;
        }
        UnitOfWork Session
        {
            get { return (UnitOfWork)items.Session; }
        }
        public StorageItem FindItem(string name)
        {
            return Session.FindObject<StorageItem>(new BinaryOperator("Url", name));
        }
        public XPCollection<StorageItem> Items
        {
            get { return items; }
        }


    }

    public class StorageItem : XPObject
    {
        string url;
        byte[] layout = null;
        public string Url
        {
            get { return url; }
            set { SetPropertyValue("Url", ref url, value); }
        }
        public byte[] Layout
        {
            get { return layout; }
            set { SetPropertyValue("Layout", ref layout, value); }
        }
        public StorageItem(Session session)
            : base(session)
        {
        }
    }
}
